DIPs/dip10ss.md

```diff
@@ -0,0 +1,284 @@
+| Status:         | Will be set by the DIP manager (e.g. "Approved" or "Rejected")  |
+
+## Abstract
+
+Allow additional meta information (attributes) to be attached with enums and functions parameters, including built-in attributes such as `deprecated`.
+
+### Reference
+
+[Preliminary discussion about Enum and Function Parameter Attributes on the NG](http://forum.dlang.org/thread/cltyrthdxkkfvuxqasqw@forum.dlang.org)
+
+[(New) Pull request implementing UDAs for function parameters](https://github.com/dlang/dmd/pull/7576)
+
+[(Old) Pull request implementing UDAs for function parameters](https://github.com/dlang/dmd/pull/4783/files)
+
+## Contents
+* [Rationale](#rationale)
+* [Description](#description)
+* [Existing Solutions](#existing-solutions)
+* [Examples](#examples)
+* [Breaking Changes and Deprecations](#breaking-changes-and-deprecations)
+* [Reviews](#reviews)
+
+## Rationale
+
+It is currently not possible to attach attributes to both enums and function parameters. This excludes a few features that can be used with almost any other symbol in D.
+
+Attributes and user-defined attributes (UDA) serve as a means to provide extra meta data for a symbol. What can be said for why attributes were included as a feature in D can also be said for why they should be extended to enums and function parameters. It is benefitial to provide extra meta data about a symbol that can be used at compile-time.
+
+The concept known as "orthogonality of language features" applies here. Attributes can be applied to almost every symbol in D. A user would expect them to also be applicable to enums and function parameters.
+
+## Description
+
+The changes made to D would be relatively minor. Most of the framework for attributes already exist and it would just be a matter of extending that into the respective symbols for enums and function parameters.
+
+The following syntaxes are being proposed to be accepted:
+
+```D
+enum SomeEnum
+{
+    // attributes declared infront of enum value
+    @(90) @MyUda deprecated("reason") someEnumValue0,
+    @(91) @MyUda deprecated("reason") someEnumValue1 = 1,
+}
+
+// attributes for enum values can be retrieved with __traits(getAttributes, ...)
+static assert(__traits(getAttributes, SomeEnum.someEnumValue0)[0] == 90);
+static assert(__traits(getAttributes, SomeEnum.someEnumValue1)[0] == 91);
+
+// for functions, attributes are allowed infront of function parameters
+void someFunction(@(93) @MyUda int someParameter)
+{
+    // can use __traits(getAttributes, ...) with the paramter from inside of function
+    static assert(__traits(getAttributes, someParamter)[0] == 93);
+}
+
+void someExternalFunction()
+{
+    // attributes can be accessed for paramters by another function
+    // through existing functionality of __parameters
+    static if(is(typeof(someFunction) PT == __parameters))
+    {
+        static assert(__traits(getAttributes, PT[0 .. 1]) == 93);
```

Comment:
wouldn't it be `PT[0] == 93` or `PT[0..1]=[93]`?

Reply:
PT is the parameter type, it was missing the `[0]` at the end of the __traits() though.